rename and change some `*_reps` functions

ThomasBreuer opened the pull request:
This is the non-controversial and non-breaking part of #3265:

Provide `complement_classes` instead of `complement_class_reps`,
`hall_subgroups` instead of `hall_subgroup_reps`,
`low_index_subgroups` instead of `low_index_subgroups_reps`.
Each of the new functions returns a vector of subgroup classes.

Also move the outdated `hall_subgroup` to `deprecations.jl`.

ThomasBreuer committed:
replace `hall_subgroup_reps` by `hall_subgroups`

ThomasBreuer committed:
replace `lowindex_subgroups_reps` by `low_index_subgroups`

ThomasBreuer committed:
replace `complement_class_reps` by `complement_classes`

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -1341,19 +1334,24 @@ together with `N` generates `G`.
 """
-function complement_class_reps(G::T, N::T) where T <: GAPGroup
-   return _as_subgroups(G, GAP.Globals.ComplementClassesRepresentatives(G.X, N.X))
+function complement_classes(G::T, N::T) where T <: GAPGroup
```
This is one of those signatures that needs to be relaxed when we introduce `SubPcGroupElem`, I guess...

ThomasBreuer replied:
Yes, and here we have also an example which is subtle in the sense that the result may be empty.
In this case, its type is that of an empty array of conjugacy classes of groups, where the type of each conjugacy class is parametrized by the types of the acting group and the representative. In the case of pc groups, the representatives will be of type `SubPcGroup`, and the acting group can be either a `PcGroup` or a `SubPcGroup`.

fingolfin commented:
But if we do this then these functions behave inconsistent with `maximal_subgroups` and `subgroups`. So it is not clear to me that this is a definite "win" -- if we decide not to change those two, then I don't think we should make the changes here?

I'll discuss this some more with @fieker later.

ThomasBreuer commented:
@fingolfin Thanks for your comment.
From my point of view, a breaking change for `maximal_subgroups*` is not a problem.
For `subgroups`, we could proceed as proposed in #3265 (return conjugacy classes of subgroups) *AND* use a different name for the `subgroups` functionality that comes from Hecke.jl.

ThomasBreuer committed:
replace `maximal_subgroup_reps` by `maximal_subgroups`

- deprecate `conjugacy_classes_maximal_subgroups`, `maximal_subgroup_reps`
- let `maximal_subgroups` return subgroup classes
- adjust calls of these functions

ThomasBreuer commented:
After the discussion of #3265, what about the names `complement_classes`, `hall_subgroup_classes`, `low_index_subgroup_classes`, `maximal_subgroup_classes`, and `subgroup_classes`?
Then the names are self-explanatory (I think), all changes are non-breaking, and also Hecke.jl's `subgroups` can be kept as is.

(If we take this solution then only `subgroups`/`subgroup_reps`/`conjugacy_classes_subgroups` needs further code changes, the rest is just string replacement.)

fingolfin commented:
That would be fine by me. 

As a variation (possibly for a future PR, `subgroups` and friends could return an iterator instead of a vector.... eg  `subgroups (G::GAPGroups) = Iterators.flatten(subgroup_classes(G))`

ThomasBreuer committed:
adjust tests involving `maximal_subgroups`

ThomasBreuer committed:
rename `conjugacy_classes_subgroups` to `subgroup_classes`

and deprecate `subgroup_reps` and `subgroups` (the latter only for `GAPGroup`s)

ThomasBreuer committed:
rename `*_subgroups` to `*_subgroup_classes`

for `*` one of `hall`, `maximal`, `low_index`

ThomasBreuer committed:
do not use deprecated functions in tests

ThomasBreuer committed:
adjust one more test using a deprecated function

ThomasBreuer committed:
adjust some doctest outputs again

fingolfin commented on docs/src/Groups/subgroups.md:
````diff
@@ -43,17 +43,13 @@ intersect(::T, V::T...) where T<:GAPGroup
 The following functions return a vector of subgroups.
 
 ```@docs
-subgroups(G::GAPGroup)
````
(I thought I'd already left this as a comment, but now it's not here -- maybe I once again forgot to press "submit"? Or maybe I left this as a comment elsewhere (and you perhaps even answered it there) -- sorry in that case, and please point me to it.)

Why is this dropping `subgroups` altogether? Isn't the plan to provide both `*subgroups` and `*subgroup_classes` commands in parallel for all four commands?

ThomasBreuer replied:
Currently the proposed changes deprecate `maximal_subgroups`; there were no `complements` and `hall_subgroups` before (just a very old, already deprecated `hall_subgroup`); and `subgroups` will remain for the methods that come from Hecke (not for `GAPGroup`s), with a different behaviour which we did not want to touch.

fingolfin replied:
But won't people still find `subgroups` and try to use it in our groups and be confused about the `method not found` error? 

So either this is renamed in Hecke (CC @thofma @fieker) or we add a `subgroups` method (which then either produces an error and tells the user to use `subgroup_classes`; or, well, it return an iterator or vector over all subgroups).

I do think that for teaching and general "easer of use" the latter (adding a proper `subgroups` method) would be best. The current approach in this PR is good for pushing people to do "the right thing" (but many people react allergic to being pushed to do the right thing...).

My instincts tell me that having something like
```julia
subgroups (G::GAPGroups) = Iterators.flatten(subgroup_classes(G))
```
or
```julia
subgroups (G::GAPGroups) = collect(Iterators.flatten(subgroup_classes(G)))
```
(resp. an optimized version of the latter) would be better. What do @thofma and @fieker think about this?

thofma replied:
Yes, I think `subgroups` should exist for the casual user and it should return subgroups. Not sure if the `Iterators.flatten(subgroup_classes(G))` does this.

ThomasBreuer replied:
O.k., we define and document `subgroups` for a `GAPGroup`, and the result is an iterator of subgroups (without embeddings), analogous to `subgroups(abelian_group([2,3,4]))`, which returns a `Hecke.SubgroupIterator`.
The docstring will contain a hint that probably calling `subgroup_classes` would be better.

Should `complements`, `hall_subgroups`, `low_index_subgroups`, `maximal_subgroups` be treated analogously?

fingolfin replied:
Yes

ThomasBreuer committed:
address comments

ThomasBreuer committed:
add missing doc. for `low_index_subgroup_classes`

Why did I not get an error in the local build?

ThomasBreuer committed:
fixes

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -756,123 +758,130 @@ function Base.rand(rng::Random.AbstractRNG, C::GroupConjClass{S,T}) where S wher
+function subgroup_classes(G::GAPGroup; order::T = ZZRingElem(-1)) where T <: IntegerUnion
   L = Vector{GapObj}(GAPWrap.ConjugacyClassesSubgroups(G.X))
+  if order != -1
+    L = [x for x in L if GAPWrap.Order(GAPWrap.Representative(x)) == order]
```
Does the `==` comparison work when `order` is a `ZZRingElem` and the group is a `GapObj` (and not an `Int`)? I would have thought not (it might be hard to add a test case for that, though... perhaps if `G` is a huge group of prime order?)

ThomasBreuer replied:
Right, this comparison is not safe, the `GAP.Obj` on the left side should better be converted to the type `T`.
Concerning  a test case: In principle this would work, but there is a bug on the GAP side, see gap-system/gap/issues/5638.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -756,123 +758,130 @@ function Base.rand(rng::Random.AbstractRNG, C::GroupConjClass{S,T}) where S wher
+  if order != -1
+    L = [x for x in L if GAPWrap.Order(GAPWrap.Representative(x)) == order]
+  end
   return [GAPGroupConjClass(G, _as_subgroup_bare(G, GAPWrap.Representative(cc)), cc) for cc in L]
```
Maybe simpler & safer (except that `order` shadows `order`, so this won't work, but you get the idea..):
```suggestion
  res = [GAPGroupConjClass(G, _as_subgroup_bare(G, GAPWrap.Representative(cc)), cc) for cc in L]
  if order != -1
    filter!(x -> order(x) == order, res)
  end
  return res
```

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -756,123 +758,130 @@ function Base.rand(rng::Random.AbstractRNG, C::GroupConjClass{S,T}) where S wher
 
-Return a vector of representatives (under conjugation) for all subgroups of `G`.
-If given, only subgroups of a certain order are returned.
+Return an iterator of all subgroups in `G`.
```
```suggestion
Return an iterator over all subgroups in `G`.
```

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -756,123 +758,130 @@ function Base.rand(rng::Random.AbstractRNG, C::GroupConjClass{S,T}) where S wher
-    conjugacy_classes_maximal_subgroups(G::Group)
+    maximal_subgroup_classes(G::Group)
 
 Return the vector of all conjugacy classes of maximal subgroups of G.
```
```suggestion
Return the vector of all conjugacy classes of maximal subgroups of `G`.
```
or
```suggestion
Return a vector of all conjugacy classes of maximal subgroups of `G`.
```

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -756,123 +758,130 @@ function Base.rand(rng::Random.AbstractRNG, C::GroupConjClass{S,T}) where S wher
+    subgroup_classes(G::GAPGroup; order::T = ZZRingElem(-1)) where T <: IntegerUnion
 
-Return the vector of all conjugacy classes of subgroups of G.
+Return the vector of all conjugacy classes of subgroups of G or,
```
```suggestion
Return the vector of all conjugacy classes of subgroups of `G` or,
```
Should it really be "the vector", not "a vector"? (Of course that's beyond the scope of this PR, but while we are at it...

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -756,123 +758,130 @@ function Base.rand(rng::Random.AbstractRNG, C::GroupConjClass{S,T}) where S wher
 
-Return a vector of representatives (under conjugation) for all maximal
-subgroups of `G`.
+Return an iterator of the maximal subgroups in `G`.
```
```suggestion
Return an iterator over the maximal subgroups in `G`.
```

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -756,123 +758,130 @@ function Base.rand(rng::Random.AbstractRNG, C::GroupConjClass{S,T}) where S wher
+"""
+    low_index_subgroups(G::Group, n::Int)
+
+Return an iterator of the subgroups of index at most `n` in `G`.
```
```suggestion
Return an iterator over the subgroups of index at most `n` in `G`.
```

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -1272,37 +1273,53 @@ up to conjugacy.
+"""
+    hall_subgroups(G::Group, P::AbstractVector{<:IntegerUnion})
+
+Return an iterator of the Hall `P`-subgroups in `G`.
```
```suggestion
Return an iterator over the Hall `P`-subgroups in `G`.
```

fingolfin commented on src/Groups/sub.jl:
```diff
@@ -295,14 +239,11 @@ i.e., those subgroups that are invariant under all automorphisms of `G`.
- Permutation group of degree 3 and order 2
- Permutation group of degree 3 and order 2
- Permutation group of degree 3 and order 2
+julia> characteristic_subgroups(symmetric_group(3))
```
Ouch, we had a bad example... good catch!

fingolfin commented on src/NumberTheory/GaloisGrp/GaloisGrp.jl:
```diff
@@ -894,7 +894,7 @@ function set_orbit(G::PermGroup, H::PermGroup)
   # https://doi.org/10.1016/j.jsc.2016.02.005
 
-  l = low_index_subgroup_reps(H, 2*degree(G)^2)
+  l = map(representative, low_index_subgroup_classes(H, 2*degree(G)^2))
```
Could also use broadcasting (I am fine with either variant)
```suggestion
  l = representative.(low_index_subgroup_classes(H, 2*degree(G)^2))
```

fingolfin commented on src/NumberTheory/GaloisGrp/Group.jl:
```diff
@@ -7,7 +7,7 @@ minimal supergroups, ie. it fails in ``C_4``.
   l = [G]
   while order(l[end]) > order(U)
-    m = maximal_subgroups(l[end])
+    m = reduce(vcat, map(collect, maximal_subgroup_classes(l[end])))
```
Should this use `maximal_subgroups` again? Possibly combined with `collect`:
```suggestion
    m = collect(maximal_subgroups(l[end]))
```

fingolfin commented:
Added "backport to 1.0" label as I need this for the book.

ThomasBreuer committed:
address comments

fingolfin committed:
Merge branch 'master' into TB_no_subgroup_reps

fingolfin commented on test/Groups/subgroups_and_cosets.jl:
```diff
@@ -388,7 +389,7 @@ end
    @test fitting_subgroup(S)==sub(S,[S([3,4,1,2]), S([4,3,2,1])])
    @test frattini_subgroup(S)==sub(S,[one(S)])
-   @test frattini_subgroup(G)[1]==intersect(maximal_subgroups(G))[1]
+   @test frattini_subgroup(G)[1]==intersect(collect(Iterators.flatten(maximal_subgroup_classes(G))))[1]
```
```suggestion
   @test frattini_subgroup(G)[1]==intersect(collect(maximal_subgroups(G)))[1]
```

fingolfin commented on test/Groups/subgroups_and_cosets.jl:
```diff
@@ -35,7 +35,7 @@
    G = symmetric_group(4)
    A = alternating_group(4)
-   L = subgroups(G)
+   L = collect(Iterators.flatten(subgroup_classes(G)))
```
```suggestion
   L = collect(subgroups(G))
```

fingolfin committed:
Apply suggestions from code review
